Address a few inconsistencies in how we process root module input variables vs. called module input variables

apparentlymart opened the pull request:
This should close #29899 along with just generally creating some better consistency in how we handle root module input variables as compared to called module input variables, to make sure we're following an equivalent workflow in each case.

The starting idea here is to make the user-provided value (prior to type conversion and validation) be attached to the graph node representing a root module variable, which corresponds with our strategy of including in a called module variable node the expression that will define that variable. We can then follow the same steps during evaluation aside from skipping the expression evaluation in the root module scenario, because root module variables always arrive as concrete values.

Another level of consistency which builds on the above is that the set of variables kept inside the `terraform.EvalContext` is _always_ the so-called "final" values, after type conversion and validation. Previously that was true for called module variables but not for root module variables, which I believe was the inconsistency that led to the bug described in #29899.

This diff is noisier than I'd like because of how much it takes to wire a new argument (the raw variable values) through to the graph builders, and the need to collect the variable values a little earlier so they're ready to pass to the graph builders, but those changes are pretty mechanical and the interesting logic lives inside the plan graph builder itself, in `NodeRootVariable`, and the shared helper functions in `eval_variable.go`. My main goal is for both `NodeRootVariable` and `NodeModuleVariable` to have the same structure where the "source value" lives inside the node itself, and the graph node uses that to prepare the "final value" to save in the `EvalContext`. `NodeRootVariable` has a constant `cty.Value` (wrapped inside `terraform.InputValue`) whereas `NodeModuleVariable` has a dynamic `hcl.Expression`, but the same general principle applies.

While here I also took the opportunity to fix a historical API wart in `EvalContext`, where `SetModuleCallArguments` was built to take a set of variable values all at once but our current caller always calls with only one at a time. That is now just 
`SetModuleCallArgument` singular, to match with the new `SetRootModuleArgument` to deal with root module variables.

alisdair commented on internal/terraform/eval_variable_test.go:
```diff
@@ -0,0 +1,426 @@
+func TestPrepareFinalInputVariableValue(t *testing.T) {
+	// This is just a concise way to define a bunch of *configs.Variable
+	// objects to use in our tests below. We're only going to decode this
+	// config, not fully evaluate it.
```
This is neat.

alisdair reviewed (approved):
This seems great to me! Just one suggestion inline for removal of probably-unused arguments.

apparentlymart commented on internal/terraform/context_plan.go:
```diff
@@ -419,11 +419,10 @@ func (c *Context) planWalk(config *configs.Config, prevRunState *states.State, r
-		InputState:         prevRunState,
-		Changes:            changes,
-		MoveResults:        moveResults,
-		RootVariableValues: rootVariables,
```
I had indeed intended to remove that entirely here, so that the variable values are used directly only during graph _construction_ and not during the walk. However, I can see now that I got a bit lost in the noise and neglected to clean up the top-level mentions of it, even though (as you noted) the deeper stuff isn't referring to it anymore. Thanks! I'll see about cleaning that up.

apparentlymart replied:
Oh, looking closer I notice that the change you suggested is valid but not for the reason I explained in my previous comment: `planWalk` _does_ include the graph construction and so it does still need the variables, but it now gets them from the `opts *PlanOpts` argument along with all of the other plan options, rather than needing a separate argument to carry them.

apparentlymart commented:
@alisdair's note about the unused value `variables` drew my attention to something I'd missed in the previous pass: we still had a call to `mergeDefaultInputVariableValues` lurking here, which was dealing with the configured default values for the root variables right up front, rather than waiting until we get into the `prepareFinalInputVariableValue` function that's shared across both root and child module variables.

As far as I can tell that extra call was _relatively_ harmless but wasn't explicitly taking into account `nullable` and so probably did still have some lurking misbehavior. In the interests of sticking with the intended architecture here and having the default-value logic live only in one central place regardless of where a variable is declared, I'm going to add another commit to remove that early default value handling in favor of `prepareFinalInputVariableValue`'s logic.

apparentlymart commented:
Hi again, @alisdair! 

Thanks to you spotting that extra redundant argument in your previous review, I noticed that we had two other layers that were still partially trying to handle default values and type conversions for root module variables, but again (just like the problem I was originally trying to address here) were doing so non-comprehensively and thus just creating a muddled, uncertain situation for any code that followed about what condition the variable values might be in.

Since my original mission here was to centralize all of the variable-final-value resolution logic into one place, I've added an additional pair of commits here which aims to remove those two extra partial variable processing attempts: one in the local backend and another in the `terraform` package itself, prior to building the graph.

I intentionally retained the perhaps-unexpected requirement that the caller into `terraform.Context.Plan` must provide a value for every declared root module variable _even if the user didn't set it_ (using `cty.NilVal` to represent that case) because I remember that much earlier we noticed and fixed some undetected bugs in the frontend where it was incorrectly skipping dealing with some variables in certain cases (such as the interactive input prompts), and so I want Terraform Core to check the consistency of the frontend's work as an aid to avoiding regressions of that sort in the future.

As part of doing that I noticed that the `prepareFinalInputVariableValue` still had the child-module-centric error message phrasings that were in the other function I had originally adapted that from, and so the second new commit I added here aims to restore a similar level of specificity to the new error messages as we'd previously been generating from the now-removed partial-pre-validation step.

jbardin commented on internal/command/jsonplan/plan.go:
```diff
@@ -203,6 +199,41 @@ func (p *plan) marshalPlanVariables(vars map[string]plans.DynamicValue, schemas
+	// three different places while child module variables were only handled
+	// during the Terraform Core graph walk.
+	//
+	// For Terraform v1.2 and later we rationalized that by having the Terraform
```
In #30312 I came to mostly the same function conclusions you did here about module variables, but opted to unify handling in the module exec node. Dealing with the default feels very much outside the graph building responsibilities of creating and connecting the nodes, and I definitely found the fake expression which was injected before any evaluation very surprising.

If that's something which can be fixed but you want to leave that for later to keep this PR from growing, I can wait for this to land and do the refactoring later.

apparentlymart replied:
I'm not exactly sure what you're referring to here, probably at least in part because I've not been looking at this PR for several weeks and so I no longer have the context loaded :grinning: , but I would generally prefer to avoid growing the scope of this even more if it seems correct enough, since it already sprawled a lot more than I wanted and is at high risk of becoming conflicted as we start other work.

I believe my thought process here was that we ought to never have included the default values in this JSON plan output at all, and that they only really ended up here as an unintended side-effect of a strange separation of concerns elsewhere. The place to look for default values for variables would be a hypothetical JSON serialization of the configuration, because default values belong to the configuration rather than to the plan. Ideally it should be possible to see from this JSON plan output whether a particular variable was set or not, and perhaps one day we'll find some way to opt-in to that (arguably) better behavior, and so this special case is here to keep this oddity isolated out here in the one system we'd be changing anyway if we were to "fix" this, rather than having this backward-compatibility shim effectively spread across multiple subsystems.

jbardin replied:
That's totally reasonable. I can follow up and see if anything from 30312 is still applicable afterwards, but the changes here also seem to fix the processing order for module variables.
